shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/UTxO.hs

```diff
@@ -163,64 +163,64 @@ txouts tx =
 
 -- | Lookup a txin for a given UTxO collection
 txinLookup ::
-  TxIn crypto ->
-  UTxO crypto ->
-  Maybe (TxOut crypto)
+  TxIn crypto v ->
+  UTxO crypto v ->
+  Maybe (TxOut crypto v)
 txinLookup txin (UTxO utxo') = Map.lookup txin utxo'
 
 -- | Verify a transaction body witness
 verifyWitVKey ::
   ( Typeable kr,
-    Crypto crypto,
-    DSignable crypto (Hash crypto (TxBody crypto))
+    CV crypto v,
+    DSignable crypto (Hash crypto (TxBody crypto v))
   ) =>
-  Hash crypto (TxBody crypto) ->
-  WitVKey crypto kr ->
+  Hash crypto (TxBody crypto v) ->
+  WitVKey crypto v kr ->
   Bool
 verifyWitVKey txbodyHash (WitVKey vkey sig) = verifySignedDSIGN vkey txbodyHash sig
 
 -- | Create a witness for transaction
 makeWitnessVKey ::
-  forall crypto kr.
-  ( Crypto crypto,
-    DSignable crypto (Hash crypto (TxBody crypto))
+  forall crypto kr v.
+  ( CV crypto v,
+    DSignable crypto (Hash crypto (TxBody crypto v))
   ) =>
-  Hash crypto (TxBody crypto) ->
+  Hash crypto (TxBody crypto v) ->
   KeyPair kr crypto ->
-  WitVKey crypto 'Witness
+  WitVKey crypto v 'Witness
 makeWitnessVKey txbodyHash keys =
   WitVKey (asWitness $ vKey keys) (signedDSIGN @crypto (sKey keys) txbodyHash)
 
 -- | Create witnesses for transaction
 makeWitnessesVKey ::
-  forall crypto kr.
-  ( Crypto crypto,
-    DSignable crypto (Hash crypto (TxBody crypto))
+  forall crypto kr v.
+  ( CV crypto v,
+    DSignable crypto (Hash crypto (TxBody crypto v))
   ) =>
-  Hash crypto (TxBody crypto) ->
+  Hash crypto (TxBody crypto v) ->
   [KeyPair kr crypto] ->
-  Set (WitVKey crypto ('Witness))
+  Set (WitVKey crypto v ('Witness))
 makeWitnessesVKey txbodyHash = Set.fromList . fmap (makeWitnessVKey txbodyHash)
 
 -- | From a list of key pairs and a set of key hashes required for a multi-sig
 -- scripts, return the set of required keys.
 makeWitnessesFromScriptKeys ::
-  ( Crypto crypto,
-    DSignable crypto (Hash crypto (TxBody crypto))
+  ( CV crypto v,
+    DSignable crypto (Hash crypto (TxBody crypto v))
   ) =>
-  Hash crypto (TxBody crypto) ->
+  Hash crypto (TxBody crypto v) ->
   Map (KeyHash kr crypto) (KeyPair kr crypto) ->
   Set (KeyHash kr crypto) ->
-  Set (WitVKey crypto ('Witness))
+  Set (WitVKey crypto v ('Witness))
 makeWitnessesFromScriptKeys txbodyHash hashKeyMap scriptHashes =
   let witKeys = Map.restrictKeys hashKeyMap scriptHashes
    in makeWitnessesVKey txbodyHash (Map.elems witKeys)
 
 -- | Determine the total balance contained in the UTxO.
-balance :: UTxO crypto -> Coin
-balance (UTxO utxo) = fromIntegral $ Map.foldl' addCoins 0 utxo
+balance :: CV crypto v => UTxO crypto v -> v
+balance (UTxO utxo) = foldr addCoins vzero utxo
   where
-    addCoins !b (TxOutCompact _ a) = a + b
+    addCoins (TxOut _ a) b = vplus a b
```

Comment:
I think we still want the real constructor `TxOutCompact`. and maybe `<>` would be more clear

Reply:
Yeah, please don't undo https://github.com/input-output-hk/cardano-ledger-specs/pull/1771. 

(Maybe you started working on this with an out-of-date checkout, one that didn't include that PR, and this change got lost while merging/rebasing? If so, we should be careful that no other such changes are in this PR.)